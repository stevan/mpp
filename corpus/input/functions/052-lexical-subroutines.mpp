# Lexical subroutine without parentheses
my sub foo {
    return 42;

# Lexical subroutine with empty parentheses
my sub bar () {
    return 100;

# Lexical subroutine with parameters
my sub add ($x, $y) {
    return $x + $y;

# Package-scoped subroutine (our)
our sub global_helper {
    return "helper";

# Package-scoped subroutine with parameters
our sub multiply ($a, $b) {
    return $a * $b;
