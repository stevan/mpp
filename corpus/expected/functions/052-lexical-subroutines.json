[
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Sub",
      "name": "foo",
      "parameters": [],
      "body": [
        {
          "type": "Return",
          "value": {
            "type": "Number",
            "value": "42"
          }
        }
      ]
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Sub",
      "name": "bar",
      "parameters": [],
      "body": [
        {
          "type": "Return",
          "value": {
            "type": "Number",
            "value": "100"
          }
        }
      ]
    }
  },
  {
    "type": "Declaration",
    "declarator": "my",
    "variable": {
      "type": "Sub",
      "name": "add",
      "parameters": [
        {
          "type": "Parameter",
          "variable": {
            "type": "Variable",
            "name": "$x"
          }
        },
        {
          "type": "Parameter",
          "variable": {
            "type": "Variable",
            "name": "$y"
          }
        }
      ],
      "body": [
        {
          "type": "Return",
          "value": {
            "type": "BinaryOp",
            "operator": "+",
            "left": {
              "type": "Variable",
              "name": "$x"
            },
            "right": {
              "type": "Variable",
              "name": "$y"
            }
          }
        }
      ]
    }
  },
  {
    "type": "Declaration",
    "declarator": "our",
    "variable": {
      "type": "Sub",
      "name": "global_helper",
      "parameters": [],
      "body": [
        {
          "type": "Return",
          "value": {
            "type": "String",
            "value": "\"helper\""
          }
        }
      ]
    }
  },
  {
    "type": "Declaration",
    "declarator": "our",
    "variable": {
      "type": "Sub",
      "name": "multiply",
      "parameters": [
        {
          "type": "Parameter",
          "variable": {
            "type": "Variable",
            "name": "$a"
          }
        },
        {
          "type": "Parameter",
          "variable": {
            "type": "Variable",
            "name": "$b"
          }
        }
      ],
      "body": [
        {
          "type": "Return",
          "value": {
            "type": "BinaryOp",
            "operator": "*",
            "left": {
              "type": "Variable",
              "name": "$a"
            },
            "right": {
              "type": "Variable",
              "name": "$b"
            }
          }
        }
      ]
    }
  }
]